content/features/featuresDeepDive/mesh/boneInternals.md

```diff
@@ -0,0 +1,159 @@
+
+Only the **bind** and **rest** matrices have a setter, `Bone.setBindMatrix()` and `Bone.setRestMatrix()` respectively. When `Bone.setBindMatrix()` is called, the **absolute bind** and **absolute inverse bind** matrices are also recalculated (and for the children of this bone too).
+
+The implementation of the **absolute bind** and **absolute inverse bind** matrix calculation (in `Bone._updateAbsoluteBindMatrices()`)) assumes that all parent bones in the hierarchy have already been processed and are up to date! So, if you're updating the **bind** matrix for several bones yourself, you need to make sure you're updating them in the right order. Another solution is to update the **bind** matrices in any order you wish, *EXCEPT* for the root bone, for which you must call `Bone.setBindMatrix()` last, as setting the **bind** matrix for a bone will calculate the **absolute** and **absolute inverse bind** matrices for that bone as well as for its children.
+
+The class supports the `rotation` and `rotationQuaternion` properties to update the orientation of the bone, but internally the orientation is stored as a quaternion. This means that if you update `rotation`, a conversion to a quaternion must take place. So, whenever possible, you should use `rotationQuaternion` to gain a tiny bit of performance.
+
+To speed up the calculation of **final** matrices for all bones in a skeleton (in `Skeleton._computeTransformMatrices`), we assume that bones are sorted in such a way (in `Skeleton.bones`) that parents come before children. This means that when we calculate the **final** matrix of a bone, we can assume that the **final** matrices of all its parent bones have already been calculated. So, if you're creating a skeleton manually, you need to make sure that the bones are in the right order in the skeleton's bone array: `Skeleton.sortBones()` can do this for you.
+
+## Understanding the **bind** matrix
+
+If you're not familiar with how bones and skeletal animations work, the **bind** matrix can be a little confusing, and you may not immediately understand its usefulness. So let's try to explain what it is and how it works.
+
+Let's take an example to support this explanation:
+
+| Mesh | Skeleton view | Skeleton hierarchy |
+|-|-|-|
+| ![Mesh](/img/features/boneInternals/character.jpg!250) | ![Skeleton view](/img/features/boneInternals/character_skeleton.jpg!250) | ![Skeleton hierarchy](/img/features/boneInternals/character_skeleton_hierarchy.jpg) |
+
+Here's the PG we'll be using throughout this section, which can also be used as an example of how to create a skeleton and bones programmatically:
+
+<Playground id="#4IUBBK#2" title="Character example" description="Simple example of creating a skeleton and bones programmatically."/>

```

Comment:
This linked PG is not using the lines view but the spheres/spurs view, should we update to #4IUBBK#7 which has lines?

Reply:
Actually, this PG has no skeleton view at all, the code is commented out. For the other PGs where I needed to display the skeleton, I only uncommented the line 81 to have the "line" rendering. I don't want the "spheres/spurs" view (I tested it, that's why there's the commented line), but it's not clear enough for my liking (it can only draw bones from current position to next bone position).